test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_core_test.clj

```diff
@@ -301,3 +311,53 @@
           "Unexpected port for second connector")
       (is (= 42 (.getSoLingerTime (second connectors)))
           "Unexpected so linger time for second connector"))))
+
+(deftest test-idle-timeout
+  (let [read-lines (fn [r]
+                     (let [sb (StringBuilder.)]
+                       (loop [l (.readLine r)]
+                         (when l
+                           (.append sb l)
+                           (.append sb "\n")
+                           ;; readLine will block until the socket is closed,
+                           ;; or will throw a SocketTimeoutException if there
+                           ;; is no data available within the SoTimeout value.
+                           (recur (.readLine r))))
+                       (.toString sb)))
+        body "Hi World\n"
+        path "/hi_world"
+        ring-handler (fn [req] {:status 200 :body body})
+        read-response (fn [client-so-timeout]
+                        (let [s (Socket. "localhost" 9000)
+                              out (PrintWriter. (.getOutputStream s) true)]
+                          (.setSoTimeout s client-so-timeout)
+                          (.println out (str "GET " path " HTTP/1.1\n"
+                                             "Host: localhost\n"
+                                             "\n"))
+                          (let [in (BufferedReader. (InputStreamReader. (.getInputStream s)))]
+                            (read-lines in))))]
+    (let [config {:webserver {:port 9000
+                              :host "localhost"
+                              :idle-timeout-milliseconds 500}}]
+      (with-test-logging
+        (with-app-with-config app
+          [jetty9-service]
+          config
+          (let [s (tk-app/get-service app :WebserverService)
+                add-ring-handler (partial add-ring-handler s)]
+            (add-ring-handler ring-handler path)
+
+            (testing "Verify that server doesn't close socket before idle timeout"
+              ;; if we set the client socket timeout lower than the server
+              ;; socket timeout, we should get a timeout exception from the
+              ;; client side while attempting to read from the socket.
+              (is (thrown-with-msg? SocketTimeoutException #"Read timed out"
+                                    (read-response 250))))
+            (testing "Verify that server closes the socket after idle timeout"
+              ;; if we set the client socket timeout higher than the server,
+              ;; then the server should close the socket after its timeout,
+              ;; which will cause our read to stop blocking and allow us to
+              ;; validate the contents of the data we read from the socket.
+              (let [resp (read-response 750)]
```

Comment:
Hmmm, yeah, I think you're right.  Since this is by default a "keep-alive" request, the server "shouldn't" close the connection right after delivering the response and so, yeah, it would seem that the client timeout would fire sooner and the test would fail.

So maybe the only potential failures might be timing-related like with the previous test.  Like if the bytes coming back from the server might be really sluggish, it could be possible for a significantly longer timeout on the server to fire while the client is still slowly receiving bytes and not yet at its limit.  Trickling a byte at a time up to 30 seconds for a localhost connection, though, seems really unlikely.

I can't think of anything to significantly improve upon the timing for this kind of test, so I'm good riding with it for now and revisiting later if we see failures in CI.